File: code_snippets/java/src/main/java/usecases/asynctasks/synctoasync/FileUploadWorkflow.java
package usecases.asynctasks.synctoasync;

import static usecases.utils.ExampleStubs.*;

import dev.restate.sdk.SharedWorkflowContext;
import dev.restate.sdk.WorkflowContext;
import dev.restate.sdk.annotation.Shared;
import dev.restate.sdk.annotation.Workflow;
import dev.restate.sdk.common.DurablePromiseKey;
import dev.restate.sdk.serde.jackson.JacksonSerdes;
import develop.workflows.Email;
import usecases.utils.URL;

// <start_here>
@Workflow
public class FileUploadWorkflow {

  private static final DurablePromiseKey<URL> URL_PROMISE =
      DurablePromiseKey.of("url", JacksonSerdes.of(URL.class));

  @Workflow
  public URL run(WorkflowContext ctx) {
    URL url = ctx.run(JacksonSerdes.of(URL.class), () -> createS3Bucket());
    ctx.run(() -> uploadData(url));

    // <mark_1>
    ctx.promiseHandle(URL_PROMISE).resolve(url);
    // </mark_1>
    return url;
  }

  @Shared
  public void getUrlViaEmail(SharedWorkflowContext ctx, Email email) {
    // <mark_1>
    URL url = ctx.promise(URL_PROMISE).awaitable().await();
    // </mark_1>
    ctx.run(() -> sendEmail(url, email));
  }
}
// <end_here>

File: code_snippets/kotlin/src/main/kotlin/usecases/asynctasks/synctoasync/FileUploadWorkflow.kt
package usecases.asynctasks.synctoasync

import dev.restate.sdk.annotation.Shared
import dev.restate.sdk.annotation.Workflow
import dev.restate.sdk.common.DurablePromiseKey
import dev.restate.sdk.kotlin.KtSerdes
import dev.restate.sdk.kotlin.SharedWorkflowContext
import dev.restate.sdk.kotlin.WorkflowContext
import dev.restate.sdk.kotlin.runBlock
import develop.workflows.Email

// <start_here>
@Workflow
class FileUploadWorkflow {

  companion object {
    private val URL_PROMISE = DurablePromiseKey.of("url", KtSerdes.json<URL>())
  }

  @Workflow
  suspend fun run(ctx: WorkflowContext): URL {
    val url: URL = ctx.runBlock { createS3Bucket() }
    ctx.runBlock { uploadData(url) }

    // <mark_1>
    ctx.promiseHandle(URL_PROMISE).resolve(url)
    // </mark_1>
    return url
  }

  @Shared
  suspend fun getUrlViaEmail(ctx: SharedWorkflowContext, email: Email) {
    // <mark_1>
    val url: URL = ctx.promise(URL_PROMISE).awaitable().await()
    // </mark_1>
    ctx.runBlock { sendEmail(url, email) }
  }
}
// <end_here>

class URL

fun createS3Bucket(): URL {
  return URL()
}

fun uploadData(url: URL) {}

fun sendEmail(url: URL, email: Email) {}
